src/main/java/net/dries007/tfc/common/blockentities/BloomeryBlockEntity.java

```diff
@@ -0,0 +1,443 @@
+            }
+
+            boolean turnOff = false;
+            while (bloomery.maxInput < bloomery.inputStacks.size())
+            {
+                turnOff = true;
+                // Structure lost one or more chimney levels
+                Helpers.spawnItem(level, bloomery.worldPosition, bloomery.inputStacks.get(0));
+                bloomery.inputStacks.remove(0);
+                bloomery.markForSync();
+            }
+            while (bloomery.maxCatalyst < bloomery.catalystStacks.size())
+            {
+                turnOff = true;
+                Helpers.spawnItem(level, bloomery.worldPosition, bloomery.catalystStacks.get(0));
+                bloomery.catalystStacks.remove(0);
+                bloomery.markForSync();
+            }
+            // Structure became compromised, unlit if needed
+            if (turnOff && state.getValue(BloomeryBlock.LIT))
+            {
+                state = state.setValue(BloomeryBlock.LIT, false);
+                level.setBlockAndUpdate(pos, state);
+            }
+            if (!BloomeryBlock.canGateStayInPlace(level, pos, direction.getAxis()))
+            {
+                // Bloomery gate (the front facing) structure became compromised
+                level.destroyBlock(pos, true);
+                return;
+            }
+
+//            if (!BloomeryBlock.isFormed(level, bloomery.getInternalBlock(), state.getValue(BloomeryBlock.FACING)) || )
+//            if (!bloomery.isInternalBlockComplete() && !bloomery.catalystStacks.isEmpty())
+//            {
+//                bloomery.dumpItems();
+//            }
+
+            int oldCatalyst = bloomery.catalystStacks.size();
+            int oldInput = bloomery.inputStacks.size();
+            bloomery.addItemsFromWorld();
+            if (oldCatalyst != bloomery.catalystStacks.size() || oldInput != bloomery.inputStacks.size())
+            {
+                bloomery.markForSync();
+            }
+            bloomery.updateMoltenBlock(state.getValue(BloomeryBlock.LIT));
+        }
+    }
+
+    protected int maxCatalyst = 0, maxInput = 0; // Helper variables, not necessary to serialize
+    protected final List<ItemStack> inputStacks = new ArrayList<>();
+    protected final List<ItemStack> catalystStacks = new ArrayList<>();
+
+    private long litTick;
+    @Nullable protected BloomeryInventory inventory;
+    @Nullable protected BloomeryRecipe cachedRecipe;
+    @Nullable protected BlockPos internalBlock, externalBlock;
+
+    public BloomeryBlockEntity(BlockPos pos, BlockState state)
+    {
+        super(TFCBlockEntities.BLOOMERY.get(), pos, state, defaultInventory(0), NAME);
+    }
+
+    @Override
+    public void loadAdditional(CompoundTag nbt)
+    {
+        CODEC.parse(NbtOps.INSTANCE, nbt.get("inputStacks")).result().ifPresent(inputStacks::addAll);
+        CODEC.parse(NbtOps.INSTANCE, nbt.get("catalystStacks")).result().ifPresent(catalystStacks::addAll);
```

Comment:
I mean I guess you can do this but I've never seen this done like that before in my life. Ask alc in discord